test/recipes/15-test_out_option.t

```diff
@@ -16,46 +16,51 @@ use OpenSSL::Test::Utils;
 
 setup("test_out_option");
 
-plan skip_all => "'-out' option tests are not available on Windows"
-    if $^O eq 'MSWin32';
-
-plan tests => 11;
-
-# The following patterns should be tested:
-#
-# path        dirname
-# /usr/       /
-# /           /
-# .           .
-# ..          .
-
-test_illegal_path('/usr/');
-test_illegal_path('/');
-test_illegal_path('./');
-test_illegal_path('../');
+# Paths that should generate failure when trying to write to them.
+# Directories are a safe bet for failure on all platforms.
+# Note that directories must end with a slash here, because of how
+# File::Spec massages them into directory specs on some platforms.
+my @failure_paths = (
+    './',
+   );
+my @success_paths = (
+    'test.pem'
+   );
 
 # Test for trying to create a file in a non-exist directory
-my @chars = ("A".."Z", "a".."z", "0".."9");
 my $rand_path = "";
-$rand_path .= $chars[rand @chars] for 1..32;
+do {
+    my @chars = ("A".."Z", "a".."z", "0".."9");
+    $rand_path .= $chars[rand @chars] for 1..32;
+} while (-d File::Spec->catdir('.', $rand_path));
 $rand_path .= "/test.pem";
 
-test_illegal_path($rand_path);
-test_legal_path('test.pem');
-unlink 'test.pem';
+push @failure_paths, $rand_path;
 
-sub test_illegal_path {
-    my $path = File::Spec->canonpath($_[0]);
-
-    my $start = time();
-    ok(!run(app([ 'openssl', 'genrsa', '-out', $path, '16384'])), "invalid output path: $path");
-    my $end = time();
-    # The above process should exit in 2 seconds if the path is not valid
-    ok($end - $start < 2, "check time consumed");
+# Specifically for mingw, the NULL device might not be what's expected.
+# For example, if cross compiled and tested on the build host, perl will
+# generate an incorrect NULL device name.
+# We might expand the exceptions...
+unless (config('target') =~ m|^mingw| && $^O ne 'msys') {
```

Comment:
Even though it's possible to test Linux cross-compiled targets with qemu-user (since they use same file system), it's not really inappropriate to simply assume that perl's devnull has no meaning to *any* cross-compiled target. Incidentally this will work out even for mingw builds performed under Linux and Cygwin. Because they are customarily performed with --cross-compile-prefix. In other words one can wonder if simple check for cross-compile prefix would be adequate. Do note "adequate", as this is really what it's about. I mean it's effectively impossible to specify condition that is guaranteed to work in all circumstances, you can only go for what has best chances to work in most cases. Of course it has to work with CIs, but as long as _access is out of picture check for cross-compile prefix should do the trick...

Reply:
Oh, so now I *should* check `config('CROSS_COMPILE')`?  'cause [back over here](https://github.com/openssl/openssl/pull/5592#pullrequestreview-103906512), you told me a different story.

...  buuuuuuut ok, you say check cross compile prefix, I will check cross compile prefix.

Reply:
> Oh, so now I should check config('CROSS_COMPILE')?

I said "one can *wonder* if it would be adequate," which is not same say "do this" :-)

> 'cause back over here, you told me a different story.

But then I also said "Grrrr! MSYS2 cheats!" I.e. by that time I didn't actually realize that msys perl would convert "/dev/null" to "nul" when starting native binary such as "mingw". I thought that it would do same as Linux and Cygwin perls, i.e. actually pass "/dev/null".

Reply:
> "Grrrr! MSYS2 cheats!"

Besides, context is different now. That's what makes it hard to nail, which can be frustrating and appear contradicting at different occasions. It's multi-variable problem with no exact solution. Change one variable, like _access call removal, and suddenly check for cross-compile prefix makes more sense than other possible options. But it's just that it makes *more* sense, not that it's absolutely "the right thing to do ever"...

Reply:
> I said "one can wonder if it would be adequate," which is not same [as to] say "do this" :-)

True that...  well, it *does* cover the mingw on Linux cases, and I've been wondering more than once what happens with other explicitly cross compiled environments.  I think this covers quite a lot of the cases we commonly test for, though...  right?